### PR TITLE
Add Windows ARM64 CI

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

The goal here is to get Windows ARM64 tests running daily rather than only with a release.

* Add an Windows ARM64 CI
  * the PR tests will require using `/azp run` [see comment](https://github.com/PowerShell/PowerShell/pull/19040#issuecomment-1406742038)
  * Several templates had to be update to choose between running in an Microsoft Hosted pool or an internal hosted pool
  * Many tests were skipped or pended because of failures.  We should try to fix those issues later.

Issues filed:

- [Console APIs fail when running in ARM64 on AzDevOps](https://github.com/PowerShell/PowerShell/issues/19045)
- [Need to setup remoting on ARM64](https://github.com/PowerShell/PowerShell/issues/19046)
- [Investigate ARM64 failures](https://github.com/PowerShell/PowerShell/issues/19047)
- [Unable to unelevate on ARM64 in azDO](https://github.com/PowerShell/PowerShell/issues/19048)
- [Tests assume shell is available](https://github.com/PowerShell/PowerShell/issues/19049)
- [tests assume homepath is set](https://github.com/PowerShell/PowerShell/issues/19050)
- [Need a way to register event provider on ARM64](https://github.com/PowerShell/PowerShell/issues/19051)
- [Get-hotfix is broken on ARM64 because native lib doesn't load](https://github.com/PowerShell/PowerShell/issues/19052)
 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: .vsts-ci/templates/ci-build.yml [hunk 1]
```diff
@@ -1,16 +1,52 @@
+# should I move the image name to the pool names variable group
```
Comment (TravisEz13) [inline .vsts-ci/templates/ci-build.yml:1]:
```suggestion
```

File: .vsts-ci/templates/ci-build.yml [hunk 2]
```diff
@@ -1,16 +1,51 @@
...
+        $powerShellPath = Join-Path -Path $env:AGENT_TEMPDIRECTORY -ChildPath 'powershell'
+        Invoke-WebRequest -Uri https://raw.githubusercontent.com/PowerShell/PowerShell/master/tools/install-powershell.ps1 -outfile ./install-powershell.ps1
+        ./install-powershell.ps1 -Destination $powerShellPath
+        $vstsCommandString = "vso[task.setvariable variable=PATH]$powerShellPath;$env:PATH"
+        Write-Host "sending " + $vstsCommandString
+        Write-Host "##$vstsCommandString"
+      }
+
+    displayName: Install PowerShell
+    condition: ne('${{ parameters.pool }}', 'windows-latest')
```
Comment (adityapatwardhan) [inline .vsts-ci/templates/ci-build.yml:44]: Remove this

Comment (TravisEz13) [reply .vsts-ci/templates/ci-build.yml:44]:
```suggestion
```

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 3]
```diff
@@ -93,6 +93,10 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
     }
 
     It "Clear-Host does not injects data into PowerShell output stream" {
+        if ([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture -eq [System.Runtime.InteropServices.Architecture]::Arm64) {
```
Comment (adityapatwardhan) [inline test/powershell/Host/ConsoleHost.Tests.ps1:96]: Replace this with a common function to skip on Window ARM64

File: test/powershell/Host/ScreenReader.Tests.ps1 [hunk 4]
```diff
@@ -44,8 +44,12 @@ Describe "Validate start of console host" -Tag CI {
     }
 
     It "PSReadLine should not be auto-loaded when screen reader status is active" -Skip:(-not $IsWindows) {
+        if ([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture -eq [System.Runtime.InteropServices.Architecture]::Arm64) {
+            Set-ItResult -Pending -Because "TBD, PSReadline loads on ARM64"
```
Comment (adityapatwardhan) [inline test/powershell/Host/ScreenReader.Tests.ps1:48]: Change TBD to "investgate"

**Review by adityapatwardhan — APPROVED** [review]

File: test/powershell/Host/ScreenReader.Tests.ps1 (hunk 4 above)
Comment (kilasuit) [reply test/powershell/Host/ScreenReader.Tests.ps1:48]: I'd prefer we kept capitalisation & correct spelling so should be `Investigate`

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 5]
```diff
@@ -1995,6 +1998,11 @@ dir -Recurse `
         It "Input '<inputStr>' should successfully complete" -TestCases $testCases -Skip:(!$IsWindows) {
             param($inputStr, $expected)
 
+            if ([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture -eq [System.Runtime.InteropServices.Architecture]::Arm64) {
+                Set-ItResult -Pending -Because "TBD"
```
Comment (kilasuit) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:2002]: as per @adityapatwardhan comment & mine Change TBD to "Investigate"

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1 [hunk 6]
```diff
@@ -145,6 +145,10 @@ Describe "Native Command Processor" -tags "Feature" {
     }
 
     It "Should not block running Windows executables" -Skip:(!$IsWindows -or !(Get-Command notepad.exe)) {
+        if ([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture -eq [System.Runtime.InteropServices.Architecture]::Arm64) {
+            Set-ItResult -Pending -Because "TBD"
```
Comment (kilasuit) [inline test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1:149]: as per @adityapatwardhan comment & mine Change TBD to "Investigate"

File: test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1 [hunk 7]
```diff
@@ -180,6 +180,10 @@ Describe "Default enablement of Experimental Features" -Tags CI {
     }
 
     It "On preview builds, Experimental Features are enabled" -Skip:(!$isPreview) {
+        if ([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture -eq [System.Runtime.InteropServices.Architecture]::Arm64) {
+            Set-ItResult -Pending -Because "TBD"
```
Comment (kilasuit) [inline test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1:184]: as per @adityapatwardhan comment & mine Change TBD to "Investigate"

**Review by kilasuit — CHANGES_REQUESTED** [review]
Other than the minor codefactor issues & minor other little nitpicks I am very glad to see this.

**Pull request merged**